### PR TITLE
Support embedded Youtube Atom in Card

### LINKED ISSUE #6662: Show video
Videos won't be supported until after the 1% launch, but once they are supported we should make sure that the "Show video" config allows toggling the video functionality on a card. I assume this means reverting to the original image if this is toggled off.

### PR DESCRIPTION
**Pull request by mxdvl**
## What does this change?

Enables support for `showVideo` from facia-tool, which allows videos to be played inline/embedded directly on fronts. Previously, it would redirect to the article itself.
    
A minimum Card width of one third or three columns is required for the YouTube atom to be embedded, similar to [the existing Frontend logic](https://github.com/guardian/frontend/blob/1c4555a9f/common/app/layout/cards/CardType.scala#L20-L23).

## Why?

Closes #6662 

## Known issues

There is currently no guarantee that the first atom will be the expected video, but it happens to be the case the vast majority of the time. See https://github.com/guardian/frontend/pull/26247 for a possible fix.

Currently, only Youtube Atoms are supported.

## Screenshots

| Frontend      | DCR before  | DCR After  |
| ------------- | ----------- | ---------- |
| ![frontend][] | ![before][] | ![after][] |

[frontend]: https://github.com/guardian/dotcom-rendering/assets/76776/67376889-4293-4f51-b4a2-1bc2c947310b
[before]: https://github.com/guardian/dotcom-rendering/assets/76776/51579c03-c273-4667-9d03-d0996cd21507
[after]: https://github.com/guardian/dotcom-rendering/assets/76776/c8316e7a-5b9d-44b6-8ea6-b5b399e1d3de

### CODE REVIEW

File: dotcom-rendering/makefile [hunk 1]
```diff
@@ -200,6 +200,7 @@ gen-schema:
 	@git add src/model/article-schema.json
 	@git add src/model/front-schema.json
 	@git add src/model/block-schema.json
+	@git add src/model/tag-front-schema.json
```
Comment (mxdvl) [inline dotcom-rendering/makefile:203]: Small refactors since #7752 introduced the new schema

File: dotcom-rendering/src/model/enhanceCards.ts [hunk 2]
```diff
@@ -182,6 +184,32 @@ const enhanceTags = (tags: FETagType[]): TagType[] => {
...
+	const asset = mediaAtom.assets.find(
+		({ version }) => version === mediaAtom.activeVersion,
+	);
+	if (asset?.platform === 'Youtube') {
+		return {
+			elementId: mediaAtom.id,
+			videoId: asset.id,
+			// Size fixed to a 5:3 ratio
+			width: 500,
+			height: 300,
```
Comment (mxdvl) [inline dotcom-rendering/src/model/enhanceCards.ts:200]: Cards do not have a concept of an intrinsic size, and calculating the size of the poster image seemed brittle.

File: dotcom-rendering/src/types/front.ts [hunk 3]
```diff
@@ -132,8 +134,17 @@ export type DCRContainerPalette =
 // TODO: These may need to be declared differently than the front types in the future
 export type DCRContainerType = FEContainerType;
 
-interface FEMediaAtoms {
+export interface FEMediaAtoms {
+	id: string;
+	activeVersion: number;
+	assets: Array<{ id: string; version: number; platform: string }>;
+	posterImage: { allImages: Image[] };
```
Comment (mxdvl) [inline dotcom-rendering/src/types/front.ts:141]: I’m not sure if any of these are optional from Frontend’s point of view…

File: dotcom-rendering/src/types/video.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+/** For displaying embedded, playable videos directly in cards */
+export type CardYoutubeVideo = {
```
Comment (mxdvl) [inline dotcom-rendering/src/types/video.ts:2]: Has anyone got a better name for it?

**Review by mxdvl — COMMENTED** [review]
Little code pointers below 👇

File: dotcom-rendering/src/components/Card/Card.tsx [hunk 5]
```diff
@@ -442,7 +444,34 @@ export const Card = ({
...
+								data-chromatic="ignore"
+								data-component="youtube-atom"
+								css={css`
+									display: block;
+									position: relative;
+									${getZIndex('card-nested-link')}
+								`}
+							>
+								<Island>
+									<YoutubeBlockComponent
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/src/components/Card/Card.tsx:459]: Implemented in #8095 because it a much bigger change.

**Review by jamesgorrie — COMMENTED** [review]
This PR looks great! Whilst it makes sense I got a bit confused.

Having two types for `CardYoutubeVideo` seemed a little leaky, especially as the transformation for either type is at different stages of the rendering pipeline (`enhance` vs `cardWrappers.onlyKeepPlayIcon`).

I wonder if we could make it more explicit of what we're doing with fewer transformations and follow some existing naming?

Here's a little pseudo code as I think I've been a little muddy in my explanation: 

```ts
type DCRCardType = {
  // ...other props...

  // assuming this flattening is done as we prefer flat
  showMainVideo: FECardType.properties.showMainVideo;
  
  // video - this is identical to the image prop
  video = decideVideo(FECardType);
}



type FrontCard = {
  // ...other props...
  showVideo: DCRCardType.showMainVideo;
  video: DCRCardType.video;

  // Maybe this isn't size, but follows the same pattern as images
  // This gets in `cardWrappers` as that's what is aware of this context
  // it is existing also an explicit property, rather than a transformation to an existing property
  // it also doesn't leak the `show-only-icon` to `CardYoutubeVideo`
  videoSize: "unplayable" : "playable";
}

// Card.tsx
const Card = ({ showVideo, video, videoSize }) => {
  if (video && showVideo) {
    if (videoSize === "playable") <VideoPlayer />
    if (videoSize === "unplayable") <PosterWithIcon />
  } else {
    // image logic
  }
}
```

File: dotcom-rendering/src/model/enhanceCards.ts [hunk 6]
```diff
@@ -182,6 +184,34 @@ const enhanceTags = (tags: FETagType[]): TagType[] => {
 	});
 };
 
+const extractVideo = (
```
Comment (jamesgorrie) [inline dotcom-rendering/src/model/enhanceCards.ts:187]: small note - should this follow the convention of `decideVideo`?

File: dotcom-rendering/src/components/Card/Card.tsx [hunk 7]
```diff
@@ -442,13 +445,44 @@ export const Card = ({
...
+										hideCaption={true}
+										expired={false}
+										role="inline"
+										stickyVideos={false}
+									/>
+								</Island>
+							</div>
 						)}
+						{image.type === 'mainMedia' &&
+							mainVideo?.type !== 'embedded-playable-video' && (
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/src/components/Card/Card.tsx:479]: I was wondering that too. Could not settle on a specific one so I’ll use your suggestion.

File: dotcom-rendering/src/types/front.ts [hunk 8]
```diff
@@ -132,8 +134,17 @@ export type DCRContainerPalette =
 // TODO: These may need to be declared differently than the front types in the future
 export type DCRContainerType = FEContainerType;
 
-interface FEMediaAtoms {
+export interface FEMediaAtoms {
+	id: string;
+	activeVersion: number;
+	assets: Array<{ id: string; version: number; platform: string }>;
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/src/types/front.ts:140]: No thanks – [not unless you enforce it with a linting rule 😉 ](https://github.com/guardian/dotcom-rendering/blob/f0259305bfa4400bca5c142859cb11a9049cabbb/dotcom-rendering/.eslintrc.js#L2-L9)

File: dotcom-rendering/src/types/front.ts [hunk 9]
```diff
@@ -132,8 +134,17 @@ export type DCRContainerPalette =
...
+	posterImage: { allImages: Image[] };
 	duration?: number;
+	source?: string;
+	title?: string;
+	// fields that we’re not interested in but are still passed…
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/src/types/front.ts:145]: I don’t know what’s the best approach here? Only validate against the stuff you’re interested in using? Should it reflect what’s actually been sent by Frontend?

This is a larger question so happy to just ignore them for now.

File: dotcom-rendering/src/components/Card/Card.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (jamesgorrie) [reply dotcom-rendering/src/components/Card/Card.tsx:479]: My reasoning is that they aren't different types, it is rather a condition set by the parent component which has extra context.

[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/src/components/Card/Card.tsx:479]: I’ve refactored `getImage` so it becomes a discriminated union which helps clarify things a lot!

File: dotcom-rendering/src/components/Card/Card.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/src/components/Card/Card.tsx:488]: Which is now the branch against which this PR is set–this refactor should happen first and then we can introduce video on Fronts.

File: dotcom-rendering/src/types/video.ts (hunk 4 above)
Comment (jamesgorrie) [reply dotcom-rendering/src/types/video.ts:2]: `Video`?

File: dotcom-rendering/src/components/Card/Card.tsx [hunk 10]
```diff
@@ -218,28 +210,65 @@ const CommentFooter = ({
...
+	type: 'video';
+	mainVideo: CardYoutubeVideo;
+};
+
+type MediaUnion =
+	| MediaSlideshow
+	| MediaAvatar
+	| MediaCrossword
+	| MediaMainMedia
+	| MediaVideo;
```
Comment (mxdvl) [inline dotcom-rendering/src/components/Card/Card.tsx:247]: Is this helping or should we simply rely on inferred return types? I might refactor and opt for conciseness.

**Review by mxdvl — COMMENTED** [review]
Excellent suggestions @jamesgorrie — I think the decision of whether to have an playable embed should indeed not be tied in the the `Video` type.

File: dotcom-rendering/src/components/Card/Card.tsx (hunk 10 above)
Comment (mxdvl) [reply dotcom-rendering/src/components/Card/Card.tsx:247]: Yep, all gone!

File: dotcom-rendering/src/types/front.ts (hunk 3 above)
Comment (mxdvl) [reply dotcom-rendering/src/types/front.ts]: I do now: https://github.com/guardian/frontend/blob/0bf69f55a/common/app/model/content/Atom.scala#L158-L169

File: dotcom-rendering/src/types/front.ts [hunk 11]
```diff
@@ -134,8 +136,28 @@ export type DCRContainerPalette =
...
+	defaultHtml: unknown;
+	assets: MediaAsset[];
+	title: string;
 	duration?: number;
+	source?: string;
+	posterImage?: { allImages: Image[] };
+	expired?: boolean;
+	activeVersion?: number;
+	/** @deprecated */
+	channelId?: unknown;
```
Comment (ioannakok) [inline dotcom-rendering/src/types/front.ts:160]: Why are `defaultHtml` and `channelId` unknown and not string?

Comment (mxdvl) [reply dotcom-rendering/src/types/front.ts:160]: Based on a [similar question](https://github.com/guardian/dotcom-rendering/pull/8075#discussion_r1247631649) by @jamesgorrie I decided to remove them, but forgot these ones.

My thinking was that we knew these fields come through, but we are not currently using them in any meaningful way. I wanted to make it clear to future travellers. I’ll just comment them out for now.

File: dotcom-rendering/src/components/Card/Card.tsx [hunk 12]
```diff
@@ -420,37 +416,67 @@ export const Card = ({
...
+									src={image.avatarUrl}
 									alt={byline ?? ''}
 									containerPalette={containerPalette}
 									format={format}
 								/>
 							</AvatarContainer>
 						)}
+						{image.type === 'video' && (
+							<div
+								data-chromatic="ignore"
```
Comment (mxdvl) [inline dotcom-rendering/src/components/Card/Card.tsx:442]: Chromatic will ignore the content of this element – ping @cemms1.

**Comment by mxdvl** [comment]
Waiting on #7972 before merging cc @abeddow91

File: dotcom-rendering/src/types/front.ts [hunk 13]
```diff
@@ -289,9 +309,7 @@ export type DCRFrontCard = {
 	byline?: string;
 	showByline?: boolean;
 	avatarUrl?: string;
-	mediaType?: MediaType;
-	mediaDuration?: number;
-	showMainVideo: boolean;
```
Comment (mxdvl) [inline dotcom-rendering/src/types/front.ts:294]: I could not find examples of when we want to hide the main video, so have removed as a simplication step.

**Review by arelra — APPROVED** [review]
A fine example of a collaborative PR!

File: dotcom-rendering/src/types/mainMedia.ts [hunk 14]
```diff
@@ -0,0 +1,28 @@
+type Media = {
+	type: 'Video' | 'Audio' | 'Gallery';
+};
+
+/** For displaying embedded, playable videos directly in cards */
+type Video = Media & {
+	type: 'Video';
+	elementId: string;
```
Comment (mxdvl) [inline dotcom-rendering/src/types/mainMedia.ts:8]: This should not have been called `elementId` but simply `id`.

Fix in #10218

**Pull request merged**